### PR TITLE
Simplify core functions logic

### PR DESCRIPTION
**Pull request by liadaram1**
- Simplify the choose course logic according to the new method with the pdf grade sheet.
- Reduce code duplication in the tests module by using a macro.
- Remove ug parser support.
- Add another test for the core function with extensive edge cases.

### CODE REVIEW

File: packages/server/src/user.rs [hunk 1]
```diff
@@ -21,42 +21,24 @@ pub struct UserDetails {
 }
 
 impl UserDetails {
-    pub fn get_mut_course_status(&mut self, id: &str) -> Option<&mut CourseStatus> {
-        for course_status in &mut self.degree_status.course_statuses.iter_mut().rev() {
+    pub fn get_course_status(&mut self, id: &str) -> Option<&CourseStatus> {
+        // returns the first course_status with the given id
+        for course_status in &mut self.degree_status.course_statuses.iter() {
```
Comment (benny-n) [inline packages/server/src/user.rs:26]: if this changes from mut course status to not mut, it shouldn't take mut self ref.

```rust
...
pub fn get_course_status(&self, id: &str) -> Option<&CourseStatus> {
    ...
    for course_status in self.degree_status.course_statuses.iter() {
    ...
```

**Review by benny-n — CHANGES_REQUESTED** [review]

**Review by benny-n — APPROVED** [review]
